### PR TITLE
Bec template fn

### LINKED ISSUE #12: Implement `create_bgc_template()`
Added in #9 but currently commented out

### PR DESCRIPTION
**Pull request by gcperk**
Hey @ateucher, here is the updated create_bgc_template.R along with testing files. I added a new helper function to assist with the testing. 
This should close #12 . 

I am also getting a fail on the create_base_vectors testing. Although this seems to be an  internet connection issues when running the tests. If I switch off my internet and the test is skipped it is all fine. Can you check if this is also failing for you? 

This is the error message when I run the test with internet on. 

```══ Failed tests ════════════════════════════════════════════════════════════════
── Error ('test-create_base_vectors.R:35:3'): create_base_vectors works with file ──
Error: No access to internet
Backtrace:
     ▆
  1. ├─PEMprepr::create_base_vectors(...) at test-create_base_vectors.R:35:2
  2. └─PEMprepr:::create_base_vectors.character(...)
  3.   ├─PEMprepr::create_base_vectors(aoi_sf, out_dir)
  4.   └─PEMprepr:::create_base_vectors.sf(aoi_sf, out_dir)
  5.     └─PEMprepr:::get_harvest(aoi, out_dir)
  6.       ├─dplyr::filter(...)
  7.       ├─dplyr::select(...)
  8.       ├─bcdata::collect(...)
  9.       ├─bcdata::select(...)
 10.       ├─bcdata::filter(...)
 11.       └─bcdata::bcdc_query_geodata("cff7b8f7-6897-444f-8c53-4bb93c7e9f8b")
── Error ('test-get_cded_dem.R:24:3'): get_cded_dem works with multiple input types ──
Error in `download.file(paste_url(url, "/", basename(md5)), quiet = TRUE, 
    destfile = md5)`: cannot open URL 'https://pub.data.gov.bc.ca/datasets/175624/93m/093m16_w.dem.zip.md5'
Backtrace:
    ▆
 1. └─PEMprepr::get_cded_dem(...) at test-get_cded_dem.R:24:2
 2.   └─bcmaps::cded_terra(aoi, ...)
 3.     └─bcmaps::cded(...)
 4.       ├─base::unlist(...)
 5.       └─base::lapply(mapsheets, get_mapsheet_tiles, dir = cded_dir, check_tiles = check_tiles)
 6.         └─bcmaps (local) FUN(X[[i]], ...)
 7.           └─utils::download.file(...)
── Error ('test-get_cded_dem.R:68:3'): res works ───────────────────────────────
Error in `download.file(paste_url(url, "/", basename(f)), quiet = TRUE, 
    destfile = f)`: cannot open URL 'https://pub.data.gov.bc.ca/datasets/175624/93m/093m13_w.dem.zip'
Backtrace:
    ▆
 1. └─PEMprepr::get_cded_dem(...) at test-get_cded_dem.R:68:2
 2.   └─bcmaps::cded_terra(aoi, ...)
 3.     └─bcmaps::cded(...)
 4.       ├─base::unlist(...)
 5.       └─base::lapply(mapsheets, get_mapsheet_tiles, dir = cded_dir, check_tiles = check_tiles)
 6.         └─bcmaps (local) FUN(X[[i]], ...)
 7.           └─utils::download.file(...)

[ FAIL 3 | WARN 12 | SKIP 0 | PASS 24 ]
Error: Test failures
Execution halted

1 error ✖ | 0 warnings ✔ | 1 note ✖```


This might need a separate issue created? 
Thanks for the review. 
Gen

### CODE REVIEW

**Comment by ateucher** [comment]
Those test failures are weird - have you tried again? Could just have been a temporary connection issue or server outage?

**Review by ateucher — APPROVED** [review]
Looks great Gen! Love the addition of the test helper, though in this case I think we can use `get_BEC()`. A couple of other suggestions, including relying on `terra::vect()` to deal with either path or object input, rather than writing our own checks.

**Comment by gcperk** [comment]
@ateucher - I tried again and did not get the error so I suspect you are correct about the intermittent server access. I have had this error before when downloading from the BC catalogue.

File: R/create_bgc_template.R [hunk 1]
```diff
@@ -1,48 +1,82 @@
 #' Create BGC template
 #'
-#' @param bec_sf sf object with bec linework ro filepath
-#' @param field field within in sf object which will be used to assign raster value
-#' @param template raster object of the size in which to model output raster not including the resoution subfolder
-#' @param outpath location in which output to be saved, default is based on fid object. Note resolution of template will be used determine folder in which output is saved
+#' @param bec An `sf` object (e.g. polygon) or path to a spatial file
```
[the remark this answers is not in the document]
Comment (gcperk) [reply R/create_bgc_template.R:3]: Nice, I can update these link types in the functions going forward

File: tests/testthat/test-create_bgc_template.R [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  skip_on_cran()
+
+  outdir <- withr::local_tempdir()
+
+  aoi_snapped <- make_test_aoi(outdir)
+  bec_test <- make_test_bec(outdir)
+  rast_temp <- create_template_raster(aoi_snapped, write_output = FALSE)
+
+  bec_rast <- create_bgc_template(
+    bec =  bec_test,
```
[the remark this answers is not in the document]
Comment (gcperk) [reply tests/testthat/test-create_bgc_template.R]: Nice! I was thinking of this, but not aware it was available within the package. Good to know!

File: R/create_bgc_template.R [hunk 3]
```diff
@@ -1,48 +1,82 @@
...
+  # template_raster <- fs::path(PEMprepr::read_fid()$dir_1020_covariates$path_rel,"25m","template.tif")
+  ## end testing lines
+
+  # read in bec file
+  if (inherits(bec, c("character"))) {
+    bec <- sf::st_read(bec, quiet = TRUE)
+
+  } else if (!inherits(bec, c("sf", "sfc"))) {
+    cli::cli_abort("{.var bec} must be an sf or an sfc object or a path to a file")
+  }
```
[the remark this answers is not in the document]
Comment (gcperk) [reply R/create_bgc_template.R:43]: Oh cool.. ok I will try that and update

File: tests/testthat/helper.R [hunk 4]
```diff
@@ -4,3 +4,20 @@ make_test_aoi <- function(outdir) {
     out_dir = fs::path(outdir, "snap")
   )
 }
+
```
[the remark this answers is not in the document]
Comment (gcperk) [reply tests/testthat/helper.R:7]: Nice - good catch.

File: tests/testthat/test-create_bgc_template.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (ateucher) [reply tests/testthat/test-create_bgc_template.R]: Yup, all internal functions are available for testing

**Pull request merged**